New IA - Automations (#5689)
TykTechnologies/tyk-docs: tyk-docs/content/api-management/automations.md

---
description: Tyk Tools that help with automating deployment and API Management operations
linkTitle: Automation Tools
tags: ["Tyk API Management", "Open Source", "Self-Managed", "Tyk Cloud", "API Gateway"]
title: Automation Tools
---

> ## Introduction

Reviewer: Any reason that we did not follow the exact structure that was discussed in [IA](https://docs.google.com/document/d/1-ngEX98p7m8I9Z5v328ASN4aQ-B8MgVcasKnzQiULZY/edit?tab=t.0#heading=h.c2iy6qzeswnu)

For example:
For `Synchronize Tyk Configurations Across Environments` section, these topics were missing `Automated Configuration Backups`, `Cross-Environment Synchronization`, `Version-Controlled API Definitions`

Similary for `Using Tyk APIs to Manage Resources` and `Automate Multi-Environment Deployments`

or Am I missing something?

- **Access to a Kubernetes cluster** (for Tyk Operator sections)
  - If you're new to Kubernetes, check out the official [Kubernetes documentation](https://kubernetes.io/docs/setup/)

- **Helm** (for installing Tyk Operator)
  - If you don't have Helm installed, follow the [official Helm installation guide](https://helm.sh/docs/intro/install/)
  - Verify your installation by running `helm version` in your terminal

- **Tyk Dashboard access and API credentials**
  - Learn how to set up the Tyk Dashboard [here](https://tyk.io/docs/tyk-dashboard/)
  - For API credentials, see our guide on [creating API tokens](https://tyk.io/docs/tyk-dashboard/managing-users/#create-an-api-token)

- **Basic knowledge of Kubernetes, YAML, and API concepts**
  - For Kubernetes, visit the [official tutorials](https://kubernetes.io/docs/tutorials/)
  - For YAML, check out this [YAML tutorial](https://yaml.org/spec/1.2/spec.html)
  - For API concepts, review our [API management basics](https://tyk.io/docs/getting-started/key-concepts/)

If you're missing any of these prerequisites, please follow the provided links to set up the necessary components before proceeding with the lifecycle automation steps.

> ## Getting Started

Reviewer: This page defines 4 sections for automation, It seems this `getting started` guide is only intended for Tyk operators & it has no relation with the other 3 sections. Can we move this part into `Automate API Management in Kubernetes Environments`

      auth_header_name: Authorization
---
apiVersion: tyk.tyk.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: httpbin
spec:
  name: Rate Limit, Quota and Throttling policy
  state: active
  active: true
  access_rights_array:
    - name: httpbin
      namespace: default
      versions:
        - Default
  quota_max: 10
  quota_renewal_rate: 60
  rate: 5
  per: 5
  throttle_interval: 2
  throttle_retry_limit: 2
EOF
```

To check that policy has been created, you can run the following command:

```console
$ kubectl get securitypolicy
NAME      AGE
httpbin   10s
```

You have successfully created the  `httpbin` security policy for your `httpbin` API. 

#### Security Policy CRD

You can use SecurityPolicy CRD to set access lists for API and versions, global usage quota, rate limits, and throttling, and also add tags and metadata:

```yaml
apiVersion: tyk.tyk.io/v1alpha1
kind: SecurityPolicy            # SecurityPolicy CRD
metadata:
  name: httpbin                 # Unique k8s name
spec:
  name: Httpbin Security Policy # Generic Name
  state: active                 # View securitypolicy_types for more info
  active: true                  # View securitypolicy_types for more info

>   access_rights_array:          # Adding APIs to the Policy. More info just below
>     - name: httpbin             # Metadata name of API
>       namespace: default
>       versions:
>         - Default               # Mandatory, Default is created automatically
>       allowed_urls:             # Path-based permissions
>         - url: /get
>           methods:
>             - GET
>   quota_max: 10
>   quota_renewal_rate: 60
>   rate: 5
>   per: 5
>   throttle_interval: 2
>   throttle_retry_limit: 2
>   tags:
>     - Hello
>     - World
>   meta_data:
>     key: value
>     hello: world
> ```
> 
> 
> Required fields in the policy:
> 
> - `name`: The name of the security policy.
> - `active`: Marks policy as active.
> - `state`: It can have value `active, draft,deny`.
> 
> Access lists for API and versions:
> 
> - `access_right_array`: The list of APIs security policy has access to.
> 
> Path-based permissions for API:
> 
> - `allowed_urls`: Restrict access per path and per method to specific portions of the API
> 
> Usage Quota fields:
> 
> - `quota_max`: The maximum number of allowed requests over a quota period.
> - `quota_renewal_rate`: Time, in seconds, after which quota will be renewed.
> 
> Rate limiting fields:
> 
> - `rate`: The number of the requests to allow per period.
> - `per`: Time in seconds.
> 
> Throttling fields:
> 
> - `throttle_interval`: Interval (in seconds) between each request retry.
> - `throttle_retry_limit`: Total requests retry number.
> 
> Tags:
> 
> - `tags`: List of tags.
> 
> Meta data:
> 
> - `meta_data`: Metadata key and values.

Reviewer: ```suggestion
```

I think we can remove it & point to the definition docs instead. Because we haven't explained the same for `ApiDefinition` CRD

  - /tyk-operator
  - /tyk-stack/tyk-operator/access-an-api
  - /tyk-stack/tyk-operator/create-an-api
  - /tyk-stack/tyk-operator/getting-started-tyk-operator
  - /tyk-stack/tyk-operator/installing-tyk-operator
  - /tyk-stack/tyk-operator/migration
  - /tyk-stack/tyk-operator/publish-an-api
  - /tyk-stack/tyk-operator/secure-an-api
  - /tyk-stack/tyk-operator/tyk-operator-reconciliation
  - /tyk-sync

---
## Introduction
Managing APIs across multiple environments can quickly become complex. Updating and overseeing multiple configurations, security policies, and deployments requires a significant amount of effort without the right tools. Tyk’s suite of automation tools simplifies this process by enabling automated control over API management tasks, helping teams ensure reliability, reduce manual errors, and maintain consistency across deployments.

In this guide, we’ll walk through the primary tools for automating API management with Tyk, including:

* **Tyk Operator for Kubernetes**: Automate API deployments within Kubernetes environments.
* **Tyk Sync**: Sync configurations across environments for consistent API management.

> * **Programmatic API Management**: Use Tyk’s APIs to automate tasks such as token management and policy updates.
> * **Multi-Environment Deployments**: Simplify deployments across various staging, testing, and production environments.

Reviewer: ```suggestion
```

Remove this.

### Benefits of Managing Certificates with Tyk Operator
- **Reduced operational overhead**: Automates the process of updating certificates when they rotate.
- **Minimized risk of API downtime**: Ensures that APIs continue to function smoothly, even when certificates are updated.
- **Improved developer experience**: Removes the need for API developers to manage certificate IDs manually.

### Examples

| Certificate Type | Supported in ApiDefinition | Supported in TykOasApiDefinition |
|------------------|-------------|---------|
| Client certifates | ✅ [Client mTLS]({{<ref "basic-config-and-security/security/mutual-tls/client-mtls#tyk-operator-classic">}}) | ✅ [Client mTLS]({{<ref "basic-config-and-security/security/mutual-tls/client-mtls#tyk-operator-oas">}}) |
| Custom domain certificates | ✅ [TLS and SSL]({{<ref "basic-config-and-security/security/tls-and-ssl#tyk-operator-classic">}}) | ✅ [TLS and SSL]({{<ref "basic-config-and-security/security/tls-and-ssl#tyk-operator-oas">}}) |
| Public keys pinning | ✅ [Certificate pinning]({{<ref "security/certificate-pinning#tyk-operator-classic">}}) | ✅ [Certificate pinning]({{<ref "security/certificate-pinning#tyk-operator-oas">}}) |

> | Upstream mTLS | ✅ [Upstream mTLS via Operator]({{<ref "basic-config-and-security/security/mutual-tls/upstream-mtls#tyk-operator-classic">}}) | ✅ [Upstream mTLS via Operator]({{<ref "basic-config-and-security/security/mutual-tls/upstream-mtls#tyk-operator-oas">}}) |

Author: Yes, heading level was wrong- fixed

> ##### Install and Configure Tyk Operator

Author: nope, fixed this thanks

When using the CRD for versioning configuration, you don't have to worry about knowing or managing the unique API IDs within Tyk. The Tyk Operator handles the actual API definition configuration behind the scenes, reducing the complexity of version management.

In case if there is original versioning information in the base API Definition, the versioning information will be kept and be merged with what is specified in CRD. If there are conflicts between the OAS API Definition and CRD, we will make use of CRD values as the final configuration. 

Tyk Operator would also protect you from accidentally deleting a version of an API that is being referenced by another API, maintaining your API integrity.

##### Configuring API Version in Tyk Classic API Definition{#tyk-classic-api}

For Tyk Classic API, versioning can be configured via `ApiDefinition` custom resource definition (CRD). See [Tyk Classic versioning]({{<ref "getting-started/key-concepts/versioning">}}) for a comprehensive example of configuring API versioning for Tyk Classic API with Tyk Operator.

> ### API Ownership

Reviewer: As per IA, this seem a sub heading of `Manage API MetaData`

| Meta Data & Tags               | ✅        | v0.1           | [Tags and Meta-data](#define-the-security-policy-manifest)        |
| Path and Method based permissions | ✅     | v0.1           | [Path based permission](#security-policy-example)        |
| Partitions                     | ✅        | v0.1           | [Partitioned policies](#security-policy-example)       |
| Per API limit                  | ✅        | v1.0           | [Per API Limit](#security-policy-example)        |
| Per-Endpoint limit             | ✅        | v1.0           | [Per Endpoint Limit](#security-policy-example)        |

### Tyk Operator changes not applied

**Problem:** Changes made through Tyk Operator are not reflected in your Tyk installation.

**Solution:**

1. Check Kubernetes events:
   ```bash
   kubectl get events --sort-by=.metadata.creationTimestamp
   ```
   This command shows recent events in your cluster, which may provide clues about why the changes weren't applied.

2. Verify Operator logs:
   ```bash
   kubectl logs -l app=tyk-operator
   ```

>    This command shows logs from the Tyk Operator pod, which may contain error messages or other useful information.

Reviewer 2: I followed this instructions but get nothing. Does it really work? 

I'd suggest rename "Reconciliation Troubleshooting" to "Tyk Operator changes not applied" because they are addressing same issue. We should use content from Reconciliation Troubleshooting because that is more operator specific.

2. Commit to Git: 
   ```
   cd configs
   git add .
   git commit -m "Update Tyk configurations"
   git push
   ```

### Can I sync multiple APIs to a single Git repository?

Yes, you can store multiple API definitions, policies, and other Tyk resources in a single Git repository. Tyk Sync and Tyk Operator can work with multiple resources in the same directory.

Your repository structure might look like this:
```
tyk-configs/
├── apis/
│   ├── api1.yaml
│   └── api2.yaml
├── policies/
│   ├── policy1.yaml
│   └── policy2.yaml
└── tyk-operator/
    └── operator-context.yaml
```

### How do I handle environment-specific configurations?

Use Tyk Operator's `OperatorContext` resource to define environment-specific variables. You can also use Kubernetes secrets and ConfigMaps to manage sensitive or environment-specific data.

Example `OperatorContext`:
```yaml
apiVersion: tyk.tyk.io/v1alpha1
kind: OperatorContext
metadata:
  name: production-context
spec:
  env:
    TYK_DB_ORGID: "prod-org-id"
    TYK_DB_APIAUTH: "prod-api-secret"
```

This YAML defines environment-specific variables for a production context, which can be referenced in your API definitions and policies.

> 

Reviewer 2: Where is it from? I don't think we have this feature.

  Latest CRD Spec Hash:  9169537376206027578
  Latest Transaction:
    Error:               unable to delete api due to security policy dependency=default/httpbin
    Status:              Failed
    Time:                2023-07-18T07:26:45Z
  Latest Tyk Spec Hash:  14558493065514264307
  linked_by_policies:
    Name:       httpbin
    Namespace:  default
```
or
```console
$ kubectl get tykapis httpbin -o json | jq .status.latestTransaction
{
  "error": "unable to delete api due to security policy dependency=default/httpbin",
  "status": "Failed",
  "time": "2023-07-18T07:26:45Z"
}
```
Instead of digging into Tyk Operator's logs, we can now diagnose this issue simply by looking at the `.status.latestTransaction` field. As `.status.latestTransaction.error` implies, the error is related to *SecurityPolicy* dependency. 

### Can I use Tyk Operator with multiple Tyk installations?

Yes, you can use Tyk Operator to manage multiple Tyk installations. You'll need to create separate `OperatorContext` resources for each installation:

```yaml
apiVersion: tyk.tyk.io/v1alpha1
kind: OperatorContext
metadata:
  name: prod-context
spec:
  env:
    TYK_MODE: pro

>     TYK_URL: http://tyk-gateway-prod:8080

Reviewer 2: ```suggestion
    TYK_URL: http://tyk-dashboard-prod
```

    TYK_AUTH: prod-secret
---
apiVersion: tyk.tyk.io/v1alpha1
kind: OperatorContext
metadata:
  name: staging-context
spec:
  env:
    TYK_MODE: pro

>     TYK_URL: http://tyk-gateway-staging:8080

Reviewer 2: ```suggestion
    TYK_URL: http://tyk-dashboard-staging
```

    TYK_AUTH: staging-secret
```

Then, you can specify which context to use in your API and Policy resources:

```yaml
apiVersion: tyk.tyk.io/v1alpha1
kind: ApiDefinition
metadata:
  name: my-api
spec:
  name: My API
  context: prod-context
  # ... other API configuration
```

### How do I roll back changes made with Tyk Sync?

To roll back changes made with Tyk Sync:

1. If you're using Git, check out the previous version of your configurations:
   ```bash
   git checkout <previous-commit-hash>
   ```

2. Use Tyk Sync to publish the previous version:
   ```bash
   tyk-sync sync -d http://dashboard:3000 -s <secret> -p ./
   ```

It's a good practice to maintain separate branches or tags for different environments to make rollbacks easier.

### Can I use Tyk Operator with non-Kubernetes Tyk installations?

While Tyk Operator is designed to work within a Kubernetes environment, you can still use it to manage non-Kubernetes Tyk installations. You'll need to:

1. Run Tyk Operator in a Kubernetes cluster.

> 2. Configure the `OperatorContext` to point to your external Tyk installation:
> 
> ```yaml
> apiVersion: tyk.tyk.io/v1alpha1
> kind: OperatorContext
> metadata:
>   name: external-tyk
> spec:
>   env:
>     TYK_MODE: pro
>     TYK_URL: http://external-tyk-gateway:8080
>     TYK_AUTH: external-secret
> ```

Reviewer 2: ```suggestion
2. Configure Tyk Operator to point to your external Tyk installation, e.g. via `tyk-operator-conf`, environment variable, or OperatorContext:

```yaml
    TYK_MODE: pro
    TYK_URL: http://external-tyk-dashboard
    TYK_AUTH: api-access-key
    TYK_ORG: org-id
```

This allows you to manage your external Tyk installation using Kubernetes resources.

### How do I migrate from Tyk Classic to Tyk OAS definitions?

To migrate from Tyk Classic to Tyk OAS definitions:

1. Export your existing API definitions using Tyk Sync:
   ```bash
   tyk-sync dump -d http://dashboard:3000 -s <secret> -t ./classic-apis
   ```

2. Use the Tyk OAS Converter tool (if available) or manually convert your Classic definitions to OAS format.

3. Update your CI/CD pipelines and Tyk Operator configurations to work with the new OAS definitions.

> 4. Gradually replace Classic definitions with OAS definitions in your Tyk installation.

Reviewer 2: @sharadregoti Where does this come from? 
@andyo-tyk are you aware of these steps?

Reviewer 3: Er.... no, we don't have such a tool. Not sure where this came from!

Reviewer 3: @sharadregoti can we get this removed from the docs please ASAP?

## Conclusion

With Tyk’s automation tools, you now have a set of options for streamlining API management, from handling deployments within Kubernetes to establishing consistency across multiple environments. By integrating these tools, you can simplify complex API workflows, maintain secure configurations, and save time through reduced manual intervention.

To continue building on what you’ve set up here, explore the following topics:

> - **Advanced Tyk API Management**: Leverage more of Tyk’s API capabilities for custom integrations and further automation possibilities. You can learn more about Tyk's custom integrations [here](/product-stack/tyk-operator/advanced-configurations/custom-plugins)

Reviewer 2: The link does not work

> - **GraphQL Support**: You can learn more about how to support GraphQL resources via Tyk Operator [here](/product-stack/tyk-operator/advanced-configurations/custom-plugins)

Reviewer 2: Wrong link. Also, it leads to 404 not found.

> - **CI/CD Integrations**: Learn how to embed Tyk automation into your CI/CD pipeline, enabling continuous deployment and reducing release cycle times.

Reviewer 2: Any link for this one?

Reviewer: I think CI/CD part has to be removed.

Reviewer 2: There's a few blogs we can point readers to
https://tyk.io/blog/a-practical-guide-using-tyk-operator-argocd-and-kustomize/
https://tyk.io/blog/gitops-enabled-api-management-in-kubernetes/
